working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1320,48 +1461,113 @@ a construct using patterns works like:
 
 The static type of a pattern `p` being matched against a value of type `M` is:
```

Comment:
Thanks for pushing on this to get me to clarify my thinking.

Concretely:

The context type schema for a pattern is used:

*   To calculate the context type schema of its superpattern. For example, in `[int x]`, the context type schema `int` for the subpattern is used to calculate `List<int>` for the outer pattern.

*   To provide a downwards inference context type for the matched value expression in pattern variable declarations and pattern-if statements.

The static type for a pattern is used:

*   To calculate the static type of its superpattern. For example,
    in `[int x]`, the type `int` for the subpattern is used to calculate `List<int>` for the outer pattern.

*   To determine if a null-check, variable, list, map, record, or extractor pattern is refutable or not. All of those potentially perform a type test and are only irrefutable if the type being tested is a supertype of the value's type.

    A consequence of this is that when the pattern is in an irrefutable context, if the pattern's type isn't a supertype of the value type, this becomes a compile error because now the pattern is considered refutable and it's an error to have a refutable pattern in an irrefutable context.

*   At runtime, to determine if a null-check, variable, list, map, record, or extractor pattern matches its type test.

So, basically, in an irrefutable context, the pattern's static type determines if you get a compile error because the pattern is trying to do a "downcast" which might fail. In an refutable context, that downcast is allowed statically and then is checked at runtime and refutes the pattern if it fails.

> Maybe if we show how it's used (being used for the subpatterns' type inference confuses me since it seems I need the types of those subpatterns to do inference on the RHS, which happens before finding the static type of the pattern match.

The information flow is a little complex. It's:

1. Calculate the context type schemas of the leaf subpatterns.
1. From those, calculate context type schemas of parent patterns.
2. Use the context type schema of the outermost pattern to do downwards inference on the value type.
3. Apply the value's type to the outermost pattern.
4. Destructure the value type and send it down into the subpatterns.
5. Calculate the types of leaf subpatterns.
6. From those, calculate the parent pattern types.

For example:

```dart
[int i, var x] = [];
```

1. First we calculate the context type schemas for `int i` and `var x`, which yields `int` and `?`.
2. From those, we infer `List<int>` as the context type schema of the outer list pattern.
3. Use that in downwards inference to infer `List<int>` for the `[]` expression.
4. The list pattern destructures that element type `int` and uses that as the matched value type to calculate the static types of its element subpatterns.
5. Calculate `int` for the type element subpatterns (so now we know `var x` has type `int`).
6. Take the GLB of those to infer `List<int>` for the list pattern.